Implement testcase rewrite: Connectivity01 and Connectivity02 + remove Basic04

pnax opened the pull request:
## Purpose

Implementation of updated testcase specification for Connectivity01 (https://github.com/zonemaster/zonemaster/pull/1097) and Connectivity02 (https://github.com/zonemaster/zonemaster/pull/1098)
Remove Basic04 (https://github.com/zonemaster/zonemaster/pull/1099).

## Context

Addresses #1136 and #1137

## Changes

* Test/Connectivity.pm
* t/Test-connectivity.t
* Remove Basic04

## How to test this PR

Should match specification and tests should pass (btw it would be nice to have more unit test with broken name servers).

mvw-afnic reviewed (changes requested):
There are some minor issues.

That being said, I’ve compared both `connectivity01` and `connectivity02` functions in their new states and the only difference I see is the transport protocol (TCP vs. UDP) in use. Maybe there is an opportunity to be seized for deduplicating the code?

mvw-afnic commented:
By the way, your comment about missing test cases reminds me of #905.

mvw-afnic commented on share/profile.json:
```diff
@@ -74,6 +74,32 @@
+            "CN01_WRONG_SOA_RECORD_UDP": "WARNING",
+            "CN02_IPV4_DISABLED": "NOTICE",
+            "CN02_IPV6_DISABLED": "NOTICE",
+            "CN02_MISSING_NS_RECORD_UDP": "WARNING",
+            "CN02_MISSING_SOA_RECORD_UDP": "WARNING",
+            "CN02_NO_RESPONSE_NS_QUERY_UDP": "WARNING",
+            "CN02_NO_RESPONSE_SOA_QUERY_UDP": "WARNING",
+            "CN02_NO_RESPONSE_UDP": "WARNING",
+            "CN02_NS_RECORD_NOT_AA_UDP": "WARNING",
+            "CN02_SOA_RECORD_NOT_AA_UDP": "WARNING",
+            "CN02_UNEXPECTED_RCODE_NS_QUERY_UDP": "WARNING",
+            "CN02_UNEXPECTED_RCODE_SOA_QUERY_UDP": "WARNING",
+            "CN02_WRONG_NS_RECORD_UDP": "WARNING",
+            "CN02_WRONG_SOA_RECORD_UDP": "WARNING",
```
These tags should end with `_TCP` instead of `_UDP`.

pnax replied:
Totally, silly copy/paste mistake.

matsduf commented:
Can implementation of https://github.com/zonemaster/zonemaster/pull/1099 be included in this PR too? It is only to remove Basic04. If so, I will merge it.

matsduf commented on lib/Zonemaster/Engine/Test/Connectivity.pm:
```diff
@@ -188,33 +294,32 @@ sub version {
 
-sub _ip_disabled_message {
-    my ( $results_array, $ns, @rrtypes ) = @_;
+sub _extract_ns {
```
There are two different tags for disabled IPv4 and IPv6. IPV4_DISABLED and IPV6_DISABLED are to be used for all test cases connecting to name servers.

matsduf commented on share/profile.json:
```diff
@@ -74,6 +74,32 @@
+            "CN01_UNEXPECTED_RCODE_SOA_QUERY_UDP": "WARNING",
+            "CN01_WRONG_NS_RECORD_UDP": "WARNING",
+            "CN01_WRONG_SOA_RECORD_UDP": "WARNING",
+            "CN02_IPV4_DISABLED": "NOTICE",
+            "CN02_IPV6_DISABLED": "NOTICE",
```
They should not exist.

pnax replied:
Indeed, the result of a copy/paste without being to careful.

pnax commented on lib/Zonemaster/Engine/Test/Connectivity.pm:
```diff
@@ -224,64 +329,109 @@ sub _ip_disabled_message {
         }
 
-        next if $ips{ $local_ns->address->short };
+        foreach my $qtype ( q{SOA NS} ) {
```
Indeed, fixed.

pnax commented on lib/Zonemaster/Engine/Test/Connectivity.pm:
```diff
@@ -224,64 +329,109 @@ sub _ip_disabled_message {
+                );
+            }
+            else {
+                my ( $soa ) = $pkt->get_records( $qtype, q{answer} );
```
Good point, I updated the variable name to `$rr`.

pnax commented on lib/Zonemaster/Engine/Test/Connectivity.pm:
```diff
@@ -224,64 +329,109 @@ sub _ip_disabled_message {
-        next if $ips{ $local_ns->address->short };
-
-        my $p = $local_ns->query( $zone->name, $query_type, { usevc => 1 } );
+        foreach my $qtype ( q{SOA NS} ) {
```
fixed

pnax commented on t/Test-connectivity.t:
```diff
@@ -10,6 +10,8 @@ BEGIN {
     use_ok( q{Zonemaster::Engine::Util} );
 }
 
+use Data::Dumper;
```
Not at all. This is now removed.

pnax commented on lib/Zonemaster/Engine/Test/Connectivity.pm:
```diff
@@ -224,64 +329,109 @@ sub _ip_disabled_message {
+                );
+            }
+            else {
+                my ( $soa ) = $pkt->get_records( $qtype, q{answer} );
```
Variable name updated to `$rr`.

pnax commented on lib/Zonemaster/Engine/Test/Connectivity.pm:
```diff
@@ -130,18 +244,10 @@ Readonly my %TAG_DESCRIPTIONS => (
     NAMESERVER_NO_TCP_53 => sub {
         __x    # CONNECTIVITY:NAMESERVER_NO_TCP_53
           'Nameserver {ns} not accessible over TCP on port 53.', @_;
     },
```
Yes, this is now removed.

pnax commented on lib/Zonemaster/Engine/Test/Connectivity.pm:
```diff
@@ -49,23 +49,39 @@ sub metadata {
             qw(
-              NAMESERVER_HAS_UDP_53
-              NAMESERVER_NO_UDP_53
-              IPV4_DISABLED
-              IPV6_DISABLED
```
I reinstated it with for Connectivity02.

tgreenx replied:
I'm not sure this is needed. I thought we decided in https://github.com/zonemaster/zonemaster/pull/1098 that Connectivity02 should not output IPVX_DISABLED messages (-- for now; [comment](https://github.com/zonemaster/zonemaster/pull/1098/files#r972747679)), and in https://github.com/zonemaster/zonemaster/pull/1097 that Connectivity01 will use CN01_IPVX_DISABLED messages instead.

If agreed then remove lines 82-83 and 415-417.

matsduf replied:
IPV6_DISABLED will be outputted every time a query is supressed because IPv6 has been disabled and is a debug message. Those are outputted in all test cases. CN01_IPV6_DISABLED is only outputted once with the list of name servers with IPv6 (if any). To be consistent, IPV6_DISABLED should always be included.

tgreenx replied:
```
              IPV4_DISABLED
              IPV6_DISABLED
```

These should be reinstated (used in line 348)

pnax replied:
Damn, I missed this. Thanks for catching it. This should be fixed now.

pnax commented on lib/Zonemaster/Engine/Test/Connectivity.pm:
```diff
@@ -224,64 +329,109 @@ sub _ip_disabled_message {
 
-        if ( _ip_disabled_message( \@results, $local_ns, $query_type ) ) {
+        if ( not $pkts_udp{SOA} and not $pkts_udp{NS} ) {
+            push @results, info( CN01_NO_RESPONSE => { ns => $ns->string } );
```
Yes, this is the same object as before.

pnax commented on lib/Zonemaster/Engine/Test/Connectivity.pm:
```diff
@@ -224,64 +329,109 @@ sub _ip_disabled_message {
-    {
+    foreach my $ns ( @$ns_keep ) {
+        my %pkts_udp = (
+            'SOA' => $ns->query( $name, q{SOA}, { usevc => 0 } ),
+            'NS'  => $ns->query( $name, q{NS}, { usevc => 0 } )
```
It should since this is the default behavior.

tgreenx commented on lib/Zonemaster/Engine/Test/Connectivity.pm:
```diff
@@ -90,6 +108,104 @@ sub metadata {
+      },
+    CN01_MISSING_NS_RECORD_UDP => sub {
+        __x    # CONNETIVITY:CN01_MISSING_NS_RECORD_UDP
+          'Nameserver {ns} reponds to a NS query with no NS records in the answer section over UDP.', @_;
```
reponds -> responds

pnax replied:
good catch (the typo is from the specification file)

tgreenx commented on lib/Zonemaster/Engine/Test/Connectivity.pm:
```diff
@@ -90,6 +108,104 @@ sub metadata {
+      },
+    CN01_MISSING_SOA_RECORD_UDP => sub {
+        __x    # CONNETIVITY:CN01_MISSING_SOA_RECORD_UDP
+          'Nameserver {ns} reponds to a SOA query with no SOA records in the answer section over UDP.', @_;
```
reponds -> responds

pnax replied:
fixed

tgreenx commented on lib/Zonemaster/Engine/Test/Connectivity.pm:
```diff
@@ -90,6 +108,104 @@ sub metadata {
 
 Readonly my %TAG_DESCRIPTIONS => (
+    CN01_IPV4_DISABLED => sub {
+        __x    # CONNETIVITY:CN01_IPV4_DISABLED
```
CONNETIVITY -> CONNECTIVITY

(everywhere thereafter, lots of occurrences)

pnax replied:
fixed

tgreenx commented on lib/Zonemaster/Engine/Test/Connectivity.pm:
```diff
@@ -90,6 +108,104 @@ sub metadata {
+
+    CN02_MISSING_NS_RECORD_TCP => sub {
+        __x    # CONNETIVITY:CN02_MISSING_NS_RECORD_TCP
+          'Nameserver {ns} reponds to a NS query with no NS records in the answer section over TCP.', @_;
```
reponds -> responds

pnax replied:
fixed

tgreenx commented on lib/Zonemaster/Engine/Test/Connectivity.pm:
```diff
@@ -90,6 +108,104 @@ sub metadata {
+      },
+    CN02_MISSING_SOA_RECORD_TCP => sub {
+        __x    # CONNETIVITY:CN02_MISSING_SOA_RECORD_TCP
+          'Nameserver {ns} reponds to a SOA query with no SOA records in the answer section over TCP.', @_;
```
reponds -> responds

pnax replied:
fixed

tgreenx commented on lib/Zonemaster/Engine/Test/Connectivity.pm:
```diff
@@ -217,71 +317,147 @@ sub _ip_disabled_message {
 }
 
+
+sub _extract_ns {
```
Suggestion: "_extract_ns" -> "_select_ns"

pnax replied:
updated

tgreenx commented on lib/Zonemaster/Engine/Test/Connectivity.pm:
```diff
@@ -217,71 +317,147 @@ sub _ip_disabled_message {
     push my @results, info( TEST_CASE_START => { testcase => (split /::/, (caller(0))[3])[-1] } );
-    my $query_type = q{SOA};
+    my $name = name( $zone );
+    my @ns = @{ Zonemaster::Engine::TestMethods->method4and5( $zone ) };
```
Suggestion (to keep it consistent): "@ns" -> "@ns_list"

pnax replied:
updated

matsduf commented:
Can implementation of https://github.com/zonemaster/zonemaster/pull/1099 be included in this PR too? It is only to remove Basic04. If so, I will merge it.

pnax commented:
> can implementation of https://github.com/zonemaster/zonemaster/pull/1099 be included in this PR too?

Yes, see https://github.com/zonemaster/zonemaster-engine/commit/5860ffa8529d38c08601692e7ec3ccbc7a657aad

matsduf commented on lib/Zonemaster/Engine/Test/Connectivity.pm:
```diff
@@ -217,71 +317,147 @@ sub _ip_disabled_message {
-    foreach
-      my $local_ns ( @{ Zonemaster::Engine::TestMethods->method4( $zone ) }, @{ Zonemaster::Engine::TestMethods->method5( $zone ) } )
-    {
+    foreach my $ns ( @$ns_keep ) {
```
Even here the following should be included to be consistent:
```
        if ( _ip_disabled_message( \@results, $ns, qw{SOA NS} ) ) {
            next;
        }
```

pnax replied:
This extra condition will always be false. This logic is already performed in `_select_ns()`.

pnax replied:
Question on the purpose here: is it the condition or the output of the DEBUG message?

matsduf replied:
The purpose is to get the same DEBUG messages per name server as for other test cases. Of course `@$ns_keep` must be replaced by a variable with all name servers.

pnax replied:
>  Of course @$ns_keep must be replaced by a variable with all name servers.

I think I misunderstood the [specification](https://github.com/zonemaster/zonemaster/blob/develop/docs/specifications/tests/Connectivity-TP/connectivity01.md) when it says: "_Extract all name servers with IPv4 [resp IPv6] address from *Name Server IP*._". I understood "extract" as remove the name server from the original set, leaving only name servers that can be usable. Did I misunderstand the meaning of this term while implementing?

matsduf replied:
The intention of the text was only to collect the name server IPv4 (IPv6) addresses for the output of the message CN01_IPV4_DISABLED (CN01_IPV6_DISABLED). It was not to change the content of the *Name Server IP* set.

pnax replied:
Ok I see. I guess this is why we have the "Special procedure requirements".

pnax replied:
I've updated the code to reflect the specification intention.

matsduf commented:
Inconsistent behavior between IPv4/v6.

```
# zonemaster-cli iis.se --test connectivity/connectivity01 --level debug --show-testcase --raw --no-ipv4 | grep IPV4
   0.18 DEBUG     CONNECTIVITY01 SKIP_IPV4_DISABLED   ns=a.root-servers.net/198.41.0.4; rrtype=NS
   0.27 DEBUG     CONNECTIVITY01 SKIP_IPV4_DISABLED   ns=a.ns.se/192.36.144.107; rrtype=NS
   0.50 NOTICE    CONNECTIVITY01 CN01_IPV4_DISABLED   ns_list=nsa.dnsnode.net/194.58.192.46;nsp.dnsnode.net/194.58.198.32;nsu.dnsnode.net/185.42.137.98
```
```
# zonemaster-cli iis.se --test connectivity/connectivity01 --level debug --show-testcase --raw --no-ipv6 | grep IPV6
   0.69 NOTICE    CONNECTIVITY01 CN01_IPV6_DISABLED   ns_list=nsa.dnsnode.net/2a01:3f1:46::53;nsp.dnsnode.net/2a01:3f1:3032::53;nsu.dnsnode.net/2a01:3f0:400::32
```

```
# zonemaster-cli iis.se --test connectivity/connectivity02 --level debug --show-testcase --raw --no-ipv4 | grep IPV4
   0.18 DEBUG     CONNECTIVITY02 SKIP_IPV4_DISABLED   ns=a.root-servers.net/198.41.0.4; rrtype=NS
   0.27 DEBUG     CONNECTIVITY02 SKIP_IPV4_DISABLED   ns=a.ns.se/192.36.144.107; rrtype=NS
   0.50 DEBUG     CONNECTIVITY02 IPV4_DISABLED   ns=nsa.dnsnode.net/194.58.192.46; rrtype=SOA
   0.50 DEBUG     CONNECTIVITY02 IPV4_DISABLED   ns=nsa.dnsnode.net/194.58.192.46; rrtype=NS
   0.51 DEBUG     CONNECTIVITY02 IPV4_DISABLED   ns=nsp.dnsnode.net/194.58.198.32; rrtype=SOA
   0.51 DEBUG     CONNECTIVITY02 IPV4_DISABLED   ns=nsp.dnsnode.net/194.58.198.32; rrtype=NS
   0.52 DEBUG     CONNECTIVITY02 IPV4_DISABLED   ns=nsu.dnsnode.net/185.42.137.98; rrtype=SOA
   0.52 DEBUG     CONNECTIVITY02 IPV4_DISABLED   ns=nsu.dnsnode.net/185.42.137.98; rrtype=NS
```

```
# zonemaster-cli iis.se --test connectivity/connectivity02 --level debug --show-testcase --raw --no-ipv6 | grep IPV6
   0.70 DEBUG     CONNECTIVITY02 IPV6_DISABLED   ns=nsa.dnsnode.net/2a01:3f1:46::53; rrtype=SOA
   0.70 DEBUG     CONNECTIVITY02 IPV6_DISABLED   ns=nsa.dnsnode.net/2a01:3f1:46::53; rrtype=NS
   0.72 DEBUG     CONNECTIVITY02 IPV6_DISABLED   ns=nsp.dnsnode.net/2a01:3f1:3032::53; rrtype=SOA
   0.72 DEBUG     CONNECTIVITY02 IPV6_DISABLED   ns=nsp.dnsnode.net/2a01:3f1:3032::53; rrtype=NS
   0.74 DEBUG     CONNECTIVITY02 IPV6_DISABLED   ns=nsu.dnsnode.net/2a01:3f0:400::32; rrtype=SOA
   0.74 DEBUG     CONNECTIVITY02 IPV6_DISABLED   ns=nsu.dnsnode.net/2a01:3f0:400::32; rrtype=NS
```

pnax commented:
> Inconsistent behavior between IPv4/v6.

This inconsistency lies within the resolution and is the result of querying A and then AAAA records (and looping in the same order). This is not related to this PR.
Consequently I don't understand the change request. Please elaborate on it.

pnax commented on lib/Zonemaster/Engine/Test/Connectivity.pm:
```diff
@@ -217,71 +317,147 @@ sub _ip_disabled_message {
-        else {
-            push @results, info( NAMESERVER_NO_TCP_53 => { ns => $local_ns->string } );
+        if ( not $pkts_tcp{SOA} and not $pkts_tcp{NS} ) {
+            push @results, info( CN02_NO_RESPONSE => { ns => $ns->string } );
```
fixed

pnax commented on lib/Zonemaster/Engine/Test/Connectivity.pm:
```diff
@@ -217,71 +317,147 @@ sub _ip_disabled_message {
 
-        if ( _ip_disabled_message( \@results, $local_ns, $query_type ) ) {
+        if ( not $pkts_udp{SOA} and not $pkts_udp{NS} ) {
+            push @results, info( CN01_NO_RESPONSE => { ns => $ns->string } );
```
fixed

pnax commented:
> why does Zonemaster output SKIP_IPV4_DISABLED when IPv4 is disabled but not SKIP_IPV6_DISABLED when IPv6 is disabled?

This is related to how Zonemaster proceeds with the resolution: A records have precedence over AAAA records. Such messages are displayed during DNS tree traversal only and not while testing the zone configuration, (see [Zone.pm (develop/4ca9539b)](https://github.com/zonemaster/zonemaster-engine/blob/4ca9539b3a09956ebcd0549a040d52aecf5fcf75/lib/Zonemaster/Engine/Zone.pm#L127-L151)).

This behavior is present in develop (grepping with some context):
```
$ git checkout develop
$ zonemaster-cli iis.se --test connectivity/connectivity01 --level debug --show-testcase --raw --no-ipv4 | grep -C 1 IPV4
   0.22 DEBUG     CONNECTIVITY01 EXTERNAL_QUERY   flags={"class":"IN"}; ip=2620:10a:80ab::150; name=se; type=SOA
   0.23 DEBUG     CONNECTIVITY01 SKIP_IPV4_DISABLED   ns=a.root-servers.net/198.41.0.4; rrtype=NS
   0.23 DEBUG     CONNECTIVITY01 EXTERNAL_QUERY   flags={"class":"IN"}; ip=2001:503:ba3e::2:30; name=se; type=NS
--
   0.31 DEBUG     CONNECTIVITY01 EXTERNAL_QUERY   flags={"class":"IN"}; ip=2a01:3f0:0:301::53; name=se; type=NS
   0.36 DEBUG     CONNECTIVITY01 SKIP_IPV4_DISABLED   ns=a.ns.se/192.36.144.107; rrtype=NS
   0.36 DEBUG     CONNECTIVITY01 EXTERNAL_QUERY   flags={"class":"IN"}; ip=2a01:3f0:0:301::53; name=iis.se; type=NS
--
   0.69 DEBUG     CONNECTIVITY01 EXTERNAL_QUERY   flags={"class":"IN"}; ip=2a01:3f1:46::53; name=iis.se; type=NS
   0.84 DEBUG     CONNECTIVITY01 IPV4_DISABLED   ns=nsa.dnsnode.net/194.58.192.46; rrtype=SOA
   0.84 DEBUG     CONNECTIVITY01 EXTERNAL_QUERY   flags={"class":"IN","usevc":0}; ip=2a01:3f1:46::53; name=iis.se; type=SOA
   0.85 DEBUG     CONNECTIVITY01 NAMESERVER_HAS_UDP_53   ns=nsa.dnsnode.net/2a01:3f1:46::53
   0.85 DEBUG     CONNECTIVITY01 IPV4_DISABLED   ns=nsp.dnsnode.net/194.58.198.32; rrtype=SOA
   0.85 DEBUG     CONNECTIVITY01 EXTERNAL_QUERY   flags={"class":"IN","usevc":0}; ip=2a01:3f1:3032::53; name=iis.se; type=SOA
   0.86 DEBUG     CONNECTIVITY01 NAMESERVER_HAS_UDP_53   ns=nsp.dnsnode.net/2a01:3f1:3032::53
   0.86 DEBUG     CONNECTIVITY01 IPV4_DISABLED   ns=nsu.dnsnode.net/185.42.137.98; rrtype=SOA
   0.86 DEBUG     CONNECTIVITY01 NAMESERVER_HAS_UDP_53   ns=nsu.dnsnode.net/2a01:3f0:400::32
   0.86 DEBUG     CONNECTIVITY01 IPV4_DISABLED   ns=nsa.dnsnode.net/194.58.192.46; rrtype=SOA
   0.86 DEBUG     CONNECTIVITY01 IPV4_DISABLED   ns=nsp.dnsnode.net/194.58.198.32; rrtype=SOA
   0.86 DEBUG     CONNECTIVITY01 IPV4_DISABLED   ns=nsu.dnsnode.net/185.42.137.98; rrtype=SOA
   0.87 DEBUG     CONNECTIVITY01 TEST_CASE_END   testcase=connectivity01
```

```
$ git checkout develop
$ zonemaster-cli iis.se --test connectivity/connectivity01 --level debug --show-testcase --raw --no-ipv6 | grep -C 1 IPV6
   0.77 DEBUG     CONNECTIVITY01 NAMESERVER_HAS_UDP_53   ns=nsa.dnsnode.net/194.58.192.46
   0.77 DEBUG     CONNECTIVITY01 IPV6_DISABLED   ns=nsa.dnsnode.net/2a01:3f1:46::53; rrtype=SOA
   0.77 DEBUG     CONNECTIVITY01 EXTERNAL_QUERY   flags={"class":"IN","usevc":0}; ip=194.58.198.32; name=iis.se; type=SOA
   0.78 DEBUG     CONNECTIVITY01 NAMESERVER_HAS_UDP_53   ns=nsp.dnsnode.net/194.58.198.32
   0.78 DEBUG     CONNECTIVITY01 IPV6_DISABLED   ns=nsp.dnsnode.net/2a01:3f1:3032::53; rrtype=SOA
   0.78 DEBUG     CONNECTIVITY01 NAMESERVER_HAS_UDP_53   ns=nsu.dnsnode.net/185.42.137.98
   0.78 DEBUG     CONNECTIVITY01 IPV6_DISABLED   ns=nsu.dnsnode.net/2a01:3f0:400::32; rrtype=SOA
   0.78 DEBUG     CONNECTIVITY01 IPV6_DISABLED   ns=nsa.dnsnode.net/2a01:3f1:46::53; rrtype=SOA
   0.79 DEBUG     CONNECTIVITY01 IPV6_DISABLED   ns=nsp.dnsnode.net/2a01:3f1:3032::53; rrtype=SOA
   0.79 DEBUG     CONNECTIVITY01 IPV6_DISABLED   ns=nsu.dnsnode.net/2a01:3f0:400::32; rrtype=SOA
   0.79 DEBUG     CONNECTIVITY01 TEST_CASE_END   testcase=connectivity01
```

pnax commented:
> Maybe there is an opportunity to be seized for deduplicating the code?

Very true. I've refactored the code in that way.

To all reviewers, hopefully I've addressed everything.

tgreenx commented on lib/Zonemaster/Engine/Test/Connectivity.pm:
```diff
@@ -221,67 +323,100 @@ sub _ip_disabled_message {
+        }
 
-        if ( _ip_disabled_message( \@results, $local_ns, $query_type ) ) {
+        my %pkts_tcp = (
```
suggestion: %pkts_tcp -> %pkts

tgreenx commented on lib/Zonemaster/Engine/Test/Connectivity.pm:
```diff
@@ -221,67 +323,100 @@ sub _ip_disabled_message {
 
-        next if $ips{ $local_ns->address->short };
+        foreach my $qtype ( qw{SOA NS} ) {
+            my $pkt = $pkts_tcp{$qtype};
```
suggestion: $pkt -> $p

tgreenx commented on lib/Zonemaster/Engine/Test/Connectivity.pm:
```diff
@@ -221,67 +323,100 @@ sub _ip_disabled_message {
-    my ( $class, $zone ) = @_;
-    push my @results, info( TEST_CASE_START => { testcase => (split /::/, (caller(0))[3])[-1] } );
-    my $query_type = q{SOA};
+sub _connectivity_loop {
```
Add an entry for that method in the documentation part of this file

tgreenx reviewed (commented):
Few comments other than that LGTM

matsduf reviewed (approved):
Looks good to me.

marc-vanderwal reviewed (approved):
Looks good to me.

matsduf commented:
@tgreenx, could you re-review this?

matsduf commented:
@pnax, please merge.

marc-vanderwal commented:
Unit tests pass. Additionally, I was able to elicit some of the more elusive messages (such as `CN01_SOA_RECORD_NOT_AA_UDP`) using IBDNS.
